### PR TITLE
Exit early on failure for local runs

### PR DESCRIPTION
**Pull request by troibe**
**What?**
With these changes the local runs fail quicker.

**Why?**
It makes it easier for the user to understand which step failed.
It only displays the last step and the tool logs only up until the error occured.
Addtionally the runtime decreases since we don't invoke unnecessary tools.

**How?**
For builtin tasks we check if any of the input dependencies successfully finished.
For non-builtin in tasks we check if all of the input dependencies successfully finished.

**Before:**
```
| INFO    | parasteps_0  | floorplan  | 0  | Tool 'openroad' found with version 'v2.0-8997' in directory '/usr/local/bin'
| ERROR   | parasteps_0  | floorplan  | 0  | Version check failed for openroad. Check installation.
| ERROR   | parasteps_0  | floorplan  | 0  | Found version v2.0-8997, did not satisfy any version specifier set >=v2.0-9413.
| ERROR   | parasteps_0  | floorplan  | 0  | Halting step 'floorplan' index '0' due to errors.
| ERROR   | parasteps_0  | physyn     | 0  | Halting step due to previous error in floorplan0
| ERROR   | parasteps_0  | physyn     | 0  | Halting step 'physyn' index '0' due to errors.
| ERROR   | parasteps_0  | place      | 0  | Halting step due to previous error in physyn0
| ERROR   | parasteps_0  | place      | 0  | Halting step 'place' index '0' due to errors.
| ERROR   | parasteps_0  | place      | 1  | Halting step due to previous error in physyn0
| ERROR   | parasteps_0  | place      | 1  | Halting step 'place' index '1' due to errors.
| INFO    | parasteps_0  | placemin   | 0  | Running builtin task 'minimum'
| ERROR   | parasteps_0  | placemin   | 0  | No inputs selected after running builtin
| ERROR   | parasteps_0  | placemin   | 0  | Halting step 'placemin' index '0' due to errors.
| ERROR   | parasteps_0  | cts        | 0  | Halting step due to previous error in placemin0
| ERROR   | parasteps_0  | cts        | 0  | Halting step 'cts' index '0' due to errors.
| ERROR   | parasteps_0  | cts        | 1  | Halting step due to previous error in placemin0
| ERROR   | parasteps_0  | cts        | 1  | Halting step 'cts' index '1' due to errors.
| INFO    | parasteps_0  | ctsmin     | 0  | Running builtin task 'minimum'
| ERROR   | parasteps_0  | ctsmin     | 0  | No inputs selected after running builtin
| ERROR   | parasteps_0  | ctsmin     | 0  | Halting step 'ctsmin' index '0' due to errors.
| ERROR   | parasteps_0  | route      | 0  | Halting step due to previous error in ctsmin0
| ERROR   | parasteps_0  | route      | 0  | Halting step 'route' index '0' due to errors.
| ERROR   | parasteps_0  | route      | 1  | Halting step due to previous error in ctsmin0
| ERROR   | parasteps_0  | route      | 1  | Halting step 'route' index '1' due to errors.
| INFO    | parasteps_0  | routemin   | 0  | Running builtin task 'minimum'
| ERROR   | parasteps_0  | routemin   | 0  | No inputs selected after running builtin
| ERROR   | parasteps_0  | routemin   | 0  | Halting step 'routemin' index '0' due to errors.
| ERROR   | parasteps_0  | dfm        | 0  | Halting step due to previous error in routemin0
| ERROR   | parasteps_0  | dfm        | 0  | Halting step 'dfm' index '0' due to errors.
| ERROR   | parasteps_0  | export     | 0  | Halting step due to previous error in dfm0
| ERROR   | parasteps_0  | export     | 0  | Halting step 'export' index '0' due to errors.
| ERROR   | parasteps_0  | export     | 1  | Halting step due to previous error in dfm0
| ERROR   | parasteps_0  | export     | 1  | Halting step 'export' index '1' due to errors.
=========================================================================== short test summary info ============================================================================
FAILED tests/examples/test_heartbeat.py::test_py - siliconcompiler._common.SiliconCompilerError: Run() failed, see previous errors.
FAILED tests/examples/test_heartbeat.py::test_cli - AssertionError: assert 1 == 0
FAILED tests/examples/test_heartbeat.py::test_parallel_all_serial - siliconcompiler._common.SiliconCompilerError: Run() failed, see previous errors.
FAILED tests/examples/test_heartbeat.py::test_parallel_steps - siliconcompiler._common.SiliconCompilerError: Run() failed, see previous errors.
=================================================================== 4 failed, 2 passed in 123.20s (0:02:03) ====================================================================
```
**After:**
```
| INFO    | parasteps_0  | floorplan  | 0  | Tool 'openroad' found with version 'v2.0-8997' in directory '/usr/local/bin'
| ERROR   | parasteps_0  | floorplan  | 0  | Version check failed for openroad. Check installation.
| ERROR   | parasteps_0  | floorplan  | 0  | Found version v2.0-8997, did not satisfy any version specifier set >=v2.0-9413.
| ERROR   | parasteps_0  | floorplan  | 0  | Halting step 'floorplan' index '0' due to errors.
=========================================================================== short test summary info ============================================================================
FAILED tests/examples/test_heartbeat.py::test_py - siliconcompiler._common.SiliconCompilerError: Run() failed, see previous errors.
FAILED tests/examples/test_heartbeat.py::test_cli - AssertionError: assert 1 == 0
FAILED tests/examples/test_heartbeat.py::test_parallel_all_serial - siliconcompiler._common.SiliconCompilerError: Run() failed, see previous errors.
FAILED tests/examples/test_heartbeat.py::test_parallel_steps - siliconcompiler._common.SiliconCompilerError: Run() failed, see previous errors.
==================================================================== 4 failed, 2 passed in 68.81s (0:01:08) ====================================================================
```

### CODE REVIEW

File: siliconcompiler/core.py [hunk 1]
```diff
@@ -4140,10 +4140,32 @@ def run(self):
                     # TODO: breakpoint logic:
                     # if task is breakpoint, then don't launch while len(running_tasks) > 0
 
+                    successful_deps = []
+                    deps_full_len = len(deps)
+                    tool, _ = self._get_tool_task(
+                        processes[task]._args[0],
+                        processes[task]._args[1])
```
Comment (nmoroze) [inline siliconcompiler/core.py:4147]: nit: I'm slightly uncomfortable with relying on these private members, it might be worth stashing the step, index tuple in a separate dictionary (or perhaps refactoring to use `(step, index)` as the actual key for all these dictionaries, if that makes sense).

Happy to keep for now and save the refactor for a separate PR, though.

File: siliconcompiler/core.py [hunk 2]
```diff
@@ -4140,10 +4140,32 @@ def run(self):
...
+
                     # Clear any tasks that have finished from dependency list.
                     for in_task in deps.copy():
                         if status[in_task] != TaskStatus.PENDING:
                             deps.remove(in_task)
+                        if status[in_task] == TaskStatus.SUCCESS:
+                            successful_deps.append(in_task)
+                        if status[in_task] == TaskStatus.ERROR:
+                            # Fail if any dependency failed for non-builtin task
+                            if not self._is_builtin(tool, task):
```
Comment (nmoroze) [inline siliconcompiler/core.py:4157]: Watch out when using `self._is_builtin(...)` - based on other uses, it expects the `task` argument to be a task name as returned by `_get_tool_task()`, but the `task` you're passing in is what we'd probably call a "node" or "node name" now (a step name + index concatenated).

It works regardless since `_is_builtin()` ignores the `task` argument for now, but it would be good to make the usage match what's expected in case that changes in the future. Part of this confusion is also related to the fact that we changed the meaning of the term "task" at some point, so it would probably be good to refactor the other usage of "task" to be "node" in a future PR.

File: siliconcompiler/core.py [hunk 3]
```diff
@@ -4140,10 +4140,32 @@ def run(self):
                     # TODO: breakpoint logic:
                     # if task is breakpoint, then don't launch while len(running_tasks) > 0
 
+                    successful_deps = []
+                    deps_full_len = len(deps)
```
Comment (nmoroze) [inline siliconcompiler/core.py:4144]: As a small cleanup suggestion, it might be possible to make each of these booleans:

```python
dep_was_successful = False
had_deps = len(deps) > 0
```
(where `deps_was_successful` gets set to True in the for loop)

and then use those directly to tighten up the `if` statement below. If you think the information as you're currently storing it might be useful for other things in the future though, then feel free to disregard.

**Review by nmoroze — CHANGES_REQUESTED** [review]
Thanks, this is a very helpful change!

The one thing I'd like to be fixed now is the task argument supplied to `_is_builtin()`. Otherwise, left a couple stylistic comments that you can address depending on your preference.

File: siliconcompiler/core.py (hunk 1 above)
Comment (troibe) [reply siliconcompiler/core.py:4147]: @gadfort @WRansohoff 
Just started refactoring but then I realised at some points the status dict does not use the `f'{step, index}'` format.
Most of these are from the server using it.
Examples I found: `status['remote_cfg']` , `status['jobhash']`, `status['slurm_constraint']`, `chip.status['slurm_account']`, `chip.status['max_fs_bytes']`, `chip.status['watchdog']`
Mixed key types in the dictionary are probably not that good of an idea.
Do you think for those setting index to 0 would make sense? i.e.`status[('jobhash', '0')]`
Or should we just completly move these out of the status dict?

In either case I think a separate PR makes more sense...

Comment (troibe) [reply siliconcompiler/core.py:4147]: SC-585

File: siliconcompiler/core.py (hunk 2 above)
Comment (troibe) [reply siliconcompiler/core.py:4157]: Ok changed that plus the meaning of node and task around it.

For the full refactor there is this ticket SC-586

File: siliconcompiler/core.py [hunk 4]
```diff
@@ -4136,21 +4136,43 @@ def run(self):
...
-                for task, deps in list(tasks_to_run.items()):
+                for node, deps in list(tasks_to_run.items()):
                     # TODO: breakpoint logic:
                     # if task is breakpoint, then don't launch while len(running_tasks) > 0
 
+                    dep_was_successful = False
+                    had_deps = len(deps) > 0
+                    tool, task = self._get_tool_task(
+                        processes[node]._args[0],
+                        processes[node]._args[1])
```
Comment (gadfort) [inline siliconcompiler/core.py:4147]: it would be nice to avoid reaching into the processing object for this information and instead track it ourselves.

**Review by gadfort — CHANGES_REQUESTED** [review]

File: siliconcompiler/core.py (hunk 4 above)
Comment (troibe) [reply siliconcompiler/core.py:4147]: Noah had similar feedback see my comment here:
https://github.com/siliconcompiler/siliconcompiler/pull/1806#discussion_r1294336467
tldr: either I add another dict or we index the status dict by step, index

File: siliconcompiler/core.py (hunk 1 above)
Comment (gadfort) [reply siliconcompiler/core.py]: okay, I think once we record the success or failure of a run in the schema, this status dict is not needed any longer.

File: siliconcompiler/core.py (hunk 4 above)
Comment (gadfort) [reply siliconcompiler/core.py:4147]: separate PR is fine

**Review by nmoroze — APPROVED** [review]
LGTM, thanks!

**Comment by gadfort** [comment]
@troibe per your request, I will not merge, once you are satisfied, you can merge

**Comment by troibe** [comment]
@gadfort You can enable merging now. The CI tests for the server passed with these changes.

**Review by gadfort — APPROVED** [review]

**Comment by gadfort** [comment]
@troibe done
